src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp

```diff
@@ -415,104 +450,191 @@ void DiscoveryDataBase::process_dispose_participant(
         pit->second.remove_participant(participant_guid.guidPrefix);
     }
 
-    if(std::find(disposals_.begin(), disposals_.end(), ch) == disposals_.end())
+    // Add entry to disposals_
+    if (std::find(disposals_.begin(), disposals_.end(), ch) == disposals_.end())
     {
         disposals_.push_back(ch);
     }
 }
 
-void DiscoveryDataBase::create_participant_from_change(
-        eprosima::fastrtps::rtps::CacheChange_t* ch)
-{
-    DiscoveryParticipantInfo part(ch);
-    participants_.insert(std::make_pair(guid_from_change(ch).guidPrefix, part));
-}
-
-void DiscoveryDataBase::create_writers_from_change(
+void DiscoveryDataBase::process_dispose_writer(
         eprosima::fastrtps::rtps::CacheChange_t* ch,
         const eprosima::fastrtps::string_255& topic_name)
 {
     const eprosima::fastrtps::rtps::GUID_t& writer_guid = guid_from_change(ch);
 
-    DiscoveryEndpointInfo tmp_writer(ch, topic_name);
-    std::pair<std::map<eprosima::fastrtps::rtps::GUID_t,DiscoveryEndpointInfo>::iterator,bool> ret;
-    ret = writers_.insert(std::make_pair(writer_guid, tmp_writer));
+    // Change DATA(w) with DATA(Uw)
+    std::map<eprosima::fastrtps::rtps::GUID_t, DiscoveryEndpointInfo>::iterator wit = writers_.find(writer_guid);
+    if (wit != writers_.end())
+    {
+        wit->second.set_disposal(ch);
+    }
 
-    if (ret.second) {
-        auto readers_it = readers_by_topic_.find(topic_name);
-        if (readers_it != readers_by_topic_.end())
+    // Update own entry participants_::writers
+    std::map<eprosima::fastrtps::rtps::GuidPrefix_t, DiscoveryParticipantInfo>::iterator pit =
+            participants_.find(writer_guid.guidPrefix);
+    if (pit != participants_.end())
+    {
+        pit->second.remove_writer(writer_guid);
+    }
+
+    //Update own entry writers_by_topic_
+    std::map<eprosima::fastrtps::string_255, std::vector<eprosima::fastrtps::rtps::GUID_t> >::iterator tit =
+            writers_by_topic_.find(topic_name);
+    if (tit != writers_by_topic_.end())
+    {
+        for (std::vector<eprosima::fastrtps::rtps::GUID_t>::iterator writer_it = tit->second.begin();
+                writer_it != tit->second.end();
+                ++writer_it)
         {
-            for (auto reader_it: readers_it->second)
+            if (*writer_it == writer_guid)
             {
-                writers_.find(writer_guid)->second.add_or_update_ack_participant(reader_it.guidPrefix);
-
-                auto rit = readers_.find(reader_it);
-                if (rit != readers_.end())
-                {
-                    rit->second.add_or_update_ack_participant(writer_guid.guidPrefix);
-                }
-
-                auto pit = participants_.find(reader_it.guidPrefix);
-                if (pit != participants_.end())
-                {
-                    pit->second.add_or_update_ack_participant(writer_guid.guidPrefix);
-                }
+                tit->second.erase(writer_it);
+                break;
             }
```

Comment:
Same as in participants, we should check here whether the list of writers in the topic is empty after we took some writers from it. Same for the reader